### PR TITLE
FAI-546: ScoreCard Model Setup Test

### PR DESCRIPTION
**Pull request by sjamboro**
See: https://issues.redhat.com/browse/FAI-546

PR contains a simple UI test suite which modify basics model attributes.

### CODE REVIEW

**Review by tomasdavidorg — APPROVED** [review]

File: packages/pmml-editor/it-tests/integration/scorecard-model.ts [hunk 1]
```diff
@@ -0,0 +1,490 @@
...
+      });
+
+    cy.ouiaType("model-setup-overview")
+      .find("[data-ouia-component-type=invalid-label]")
+      .should(($label) => {
+        expect($label).to.have.length(1);
+        expect($label[0]).to.have.text("Baseline score:\u00A0Missing");
+      });
+
+    /**
```
Comment (tomasdavidorg) [inline packages/pmml-editor/it-tests/integration/scorecard-model.ts]: Just a comment: Why do you use JSDoc comments `/**` instead of regular multi line comments `/*`?

Comment (sjamboro) [reply packages/pmml-editor/it-tests/integration/scorecard-model.ts]: Thanks, I should have used `/*` instead of `/**`. At the end I have decided to remove this comment and mention it in JS Doc comment of whole test suite.

**Comment by sjamboro** [comment]
@manstis @tiagobento @ederign can you review and merge this PR? All checks have passed.

File: packages/pmml-editor/it-tests/integration/scorecard-model.ts [hunk 2]
```diff
@@ -0,0 +1,497 @@
...
+        expect($label[1]).to.have.text("Function:\u00A0regression");
+        expect($label[2]).to.have.text("Use Reason Codes:\u00A0Yes");
+        expect($label[3]).to.have.text("Reason Code Algorithm:\u00A0pointsBelow");
+        /*
+         * The label contains different string when is invalid:
+         * - invalid label - "Baseline score:"
+         * - valid label - "Baseline Score:"
+         * TODO: reach an agreement about functionality.
+         */
+        expect($label[4]).to.have.text("Baseline Score:\u00A0-15.5");
```
Comment (sjamboro) [inline packages/pmml-editor/it-tests/integration/scorecard-model.ts:277]: @manstis: Can I correct a text of invalid label to "Baseline Score"? (to unify invalid and valid text of labels) It seems that other labels contain capital characters.

File: packages/pmml-editor/it-tests/integration/scorecard-model.ts [hunk 3]
```diff
@@ -0,0 +1,497 @@
...
+      .should(($label) => {
+        expect($label).to.have.length(6);
+        expect($label[0]).to.have.text("Is Scorable:\u00A0Yes");
+        expect($label[1]).to.have.text("Function:\u00A0regression");
+        /*
+         * This attribute is not required by PMML definition.
+         * If its value is the empty string then the particular label should be removed.
+         * TODO: reach an agreement about functionality.
+         */
+        expect($label[2]).to.have.text("Algorithm:\u00A0");
```
Comment (sjamboro) [inline packages/pmml-editor/it-tests/integration/scorecard-model.ts:351]: @manstis: What is the expected behavior of the "Algorithm" attribute? Other labels disappear when its text is empty.

**Comment by tiagobento** [comment]
Is this good to merge? I see there are comments unresolved.

File: packages/pmml-editor/it-tests/integration/scorecard-model.ts (hunk 2 above)
Comment (manstis) [reply packages/pmml-editor/it-tests/integration/scorecard-model.ts:277]: Hi @sjamboro Yes you can do this as part of this PR, if you like.

The "Characteristics" section of the editor should also probably be corrected as it only capitalises the first letter.

Screenshot from 2021-09-02 10-07-49

I had a quick check of the "Data Dictionary", "Mining Schema" and "Outputs". They seem to capitalise each word too, but there might be some edge cases you can catch and also correct.

Of course, none of these corrections are really in scope of this PR.. so feel free to make a JIRA for me instead if you prefer.

File: packages/pmml-editor/it-tests/integration/scorecard-model.ts (hunk 3 above)
Comment (manstis) [reply packages/pmml-editor/it-tests/integration/scorecard-model.ts:351]: Hi @sjamboro the _expected_ behaviour is to be consistent with other optional labels; i.e. if the property has no value then do not render it when in "view" mode (i.e. we don't show the coloured Pattern Fly Label component). When in "edit" mode we always show a UI component.

File: packages/pmml-editor/it-tests/integration/scorecard-model.ts (hunk 2 above)
Comment (sjamboro) [reply packages/pmml-editor/it-tests/integration/scorecard-model.ts:277]: @manstis ok, I am going to correct labels in model setup here. I will focus on these issues during cypress test preparation and unify all labels.

File: packages/pmml-editor/it-tests/integration/scorecard-model.ts (hunk 3 above)
Comment (sjamboro) [reply packages/pmml-editor/it-tests/integration/scorecard-model.ts:351]: @manstis so Attribute label has to be fix. Can I correct it here?

Comment (manstis) [reply packages/pmml-editor/it-tests/integration/scorecard-model.ts:351]: @sjamboro The editor looks OK (on `kogito-group/kogito-tooling/main`).. see [here](https://user-images.githubusercontent.com/497544/131816806-ff5393b1-93d4-43da-bd71-c32e7ae89151.png). The "Algorithm" attribute is not shown when in "view" mode. Is the new test finding it somewhere?! Let me checkout this branch and check.

Comment (manstis) [reply packages/pmml-editor/it-tests/integration/scorecard-model.ts:351]: @sjamboro OK, so I can replicate a "Algorithm" label with no text.. enter a value in the edit mode and then delete it. The value becomes an empty `String` and no longer `undefined`. We only check for if the value is `undefined` when rendering the component and hence it is rendered along with an empty `String` as it's value.

There are probably other places we write back an empty `String` to the model so a new JIRA to fix it across the whole editor is probably the best solution (and this test can be fixed then too).. or fix here. I don't mind. Your choice!

Comment (manstis) [reply packages/pmml-editor/it-tests/integration/scorecard-model.ts:351]: This specific case can either be fixed [here](https://github.com/kiegroup/kogito-tooling/blob/main/packages/pmml-editor/src/editor/reducers/ScorecardReducer.ts#L88) (by setting to `undefined` if an empty `String`) or [here](https://github.com/kiegroup/kogito-tooling/blob/main/packages/pmml-editor/src/editor/components/EditorScorecard/organisms/CorePropertiesTable.tsx#L268) (by setting the value passed to `setAlgorithmName(..)` to `undefined` if `e === undefined`).

Comment (manstis) [reply packages/pmml-editor/it-tests/integration/scorecard-model.ts:351]: @sjamboro Oh, BTW, please let me know what approach you decide to do (fix on this PR or a new JIRA etc) so I can pick up and fix if and as required. I won't make a JIRA until I hear what you want to do to avoid us duplicating it.

**Comment by manstis** [comment]
@tiagobento Please do not merge at the moment.

File: packages/pmml-editor/it-tests/integration/scorecard-model.ts (hunk 2 above)
Comment (sjamboro) [reply packages/pmml-editor/it-tests/integration/scorecard-model.ts:277]: @manstis please check it. I have corrected the invalid label to "Baseline Score" and removed TODO comment. All test suites passed on my laptop.

File: packages/pmml-editor/it-tests/integration/scorecard-model.ts [hunk 4]
```diff
@@ -0,0 +1,491 @@
...
```
Comment (sjamboro) [inline packages/pmml-editor/it-tests/integration/scorecard-model.ts:351]: @manstis You have better knowledge of code. Please create new JIRA where you fix it. There is suggestion how you could fix the "Attribute 'Algorithm' - set and unset" test when new behaviour will be implemented:
```suggestion
        expect($label).to.have.length(5);
        expect($label[0]).to.have.text("Is Scorable:\u00A0Yes");
        expect($label[1]).to.have.text("Function:\u00A0regression");
        expect($label[2]).to.have.text("Use Reason Codes:\u00A0Yes");
        expect($label[3]).to.have.text("Reason Code Algorithm:\u00A0pointsBelow");
        expect($label[4]).to.have.text("Baseline Method:\u00A0other");
```

Comment (manstis) [reply packages/pmml-editor/it-tests/integration/scorecard-model.ts:351]: OK, see https://issues.redhat.com/browse/FAI-558. I'll work on it next week.

**Review by manstis — APPROVED** [review]
Thank-you for your patience @sjamboro.

**Comment by manstis** [comment]
@ederign @tiagobento This is OK to merge whenever you like now.
